// @sentry/server-utils/build/esm/tracing-channel.js
import { _INTERNAL_getTracingChannelBinding, debug, captureException, SPAN_STATUS_ERROR } from '@sentry/core';
import { DEBUG_BUILD } from './debug-build.js';

const NOOP = () => {
};
function bindTracingChannelToSpan(channel, getSpan, opts) {
  const handle = bindSpanToChannelStore(channel, getSpan);
  const beforeSpanEnd = opts?.beforeSpanEnd;
  const getErrorHint = (e) => {
    if (typeof opts?.captureError === "function") {
      return opts.captureError(e);
    }
    return {
      mechanism: {
        type: "auto.diagnostic_channels.bind_span",
        handled: false
      }
    };
  };
  const subscribers = {
    start: NOOP,
    asyncStart: NOOP,
    end(data) {
      if ("error" in data || "result" in data) {
        endBoundSpan(data, beforeSpanEnd);
      }
    },
    error(data) {
      const span = data._sentrySpan;
      if (!span) {
        return;
      }
      if (opts?.captureError) {
        captureException(data.error, getErrorHint(data.error));
      }
      span.setStatus({ code: SPAN_STATUS_ERROR, message: getErrorMessage(data.error) });
    },
    asyncEnd(data) {
      endBoundSpan(data, beforeSpanEnd);
    }
  };
  handle.channel.subscribe(subscribers);
  return {
    channel: handle.channel,
    unbind: () => {
      handle.channel.unsubscribe(subscribers);
      handle.unbind();
    }
  };
}
function bindSpanToChannelStore(channel, getSpan) {
  const binding = _INTERNAL_getTracingChannelBinding();
  if (!binding) {
    DEBUG_BUILD && debug.log("[TracingChannel] Could not access async context binding.");
    return {
      channel,
      unbind: NOOP
    };
  }
  const asyncLocalStorage = binding.asyncLocalStorage;
  channel.start.bindStore(asyncLocalStorage, (data) => {
    const span = getSpan(data);
    if (!span) {
      return asyncLocalStorage.getStore();
    }
    data._sentrySpan = span;
    return binding.getStoreWithActiveSpan(span);
  });
  return {
    channel,
    unbind: () => {
      channel.start.unbindStore(asyncLocalStorage);
    }
  };
}
function endBoundSpan(data, beforeSpanEnd) {
  const span = data._sentrySpan;
  if (!span) {
    return;
  }
  beforeSpanEnd?.(span, data);
  span.end();
}
function getErrorMessage(error) {
  if (error && typeof error === "object" && "message" in error && typeof error.message === "string") {
    return error.message;
  }
  return String(error);
}

export { bindTracingChannelToSpan };
